python/ray/data/datasource/partitioning.py

```diff
@@ -0,0 +1,378 @@
+import posixpath
+from enum import Enum
+from typing import (
+    Optional,
+    List,
+    Callable,
+    Dict,
+    TYPE_CHECKING,
+)
+
+if TYPE_CHECKING:
+    import pyarrow
+
+from ray.util.annotations import DeveloperAPI
+
+
+@DeveloperAPI
+class PartitionStyle(str, Enum):
+    """Supported dataset partition styles.
+
+    Inherits from `str` to simplify plain text serialization/deserialization.
+
+    Examples:
+        >>> # Serialize to JSON text.
+        >>> json.dumps(PartitionStyle.HIVE)  # "hive"
+
+        >>> # Deserialize from JSON text.
+        >>> PartitionStyle(json.loads('"hive"'))  # PartitionStyle.HIVE
+    """
+
+    HIVE = "hive"
+    DIRECTORY = "dir"
+
+
+@DeveloperAPI
+class PathPartitionBase:
+    """Path-based partition formats embed all partition keys and values directly in
+    their dataset file paths.
+
+    Base class for path-based partition formats.
+    """
+
+    def __init__(
+        self,
+        style: PartitionStyle,
+        base_dir: Optional[str],
+        field_names: Optional[List[str]],
+    ):
+        """Creates a new path-based dataset partition scheme.
+
+        Args:
+            style: The partition style - may be either HIVE or DIRECTORY.
+            base_dir: "/"-delimited base directory that all partitioned paths should
+                exist under (exclusive).
+            field_names: The partition key field names (i.e. column names for tabular
+                datasets). Required when parsing DIRECTORY partitioned paths or
+                generating HIVE partitioned paths.
+        """
+        self._style = style
+        self._base_dir = base_dir or ""
+        self._normalized_base_dir = None
+        self._field_names = field_names
+
+    @property
+    def style(self) -> PartitionStyle:
+        """Gets the path partitioning style."""
+        return self._style
+
+    @property
+    def base_dir(self) -> str:
+        """Gets the base directory."""
+        return self._base_dir
+
+    @property
+    def normalized_base_dir(self) -> Optional[str]:
+        """Returns the base directory normalized for compatibility with a filesystem.
+
+        The normalized base directory is undefined by default, then lazily resolved
+        or updated by any operation that takes a filesystem as input.
+        """
+        return self._normalized_base_dir
+
+    @property
+    def field_names(self) -> Optional[List[str]]:
+        """Gets the partition key field names."""
+        return self._field_names
+
+    def _normalize_base_dir(self, filesystem: "pyarrow.fs.FileSystem"):
```

Comment:
Could it make more sense to have `filesystem` in the constructor?
It looks a path or partition is always in the context of a filesystem. There is no need to use the same PathPartitionBase object for different filesystems?

Reply:
Agreed - I've moved the filesystem to the constructor. We just have to make sure that `FileBasedDatasource` and `PathPartitionScheme` keep their filesystem resolution methods in-sync going forward. Passing the filesystem in post-construction made more sense in an earlier iteration when filesystem resolution wasn't built into this class.